apps/src/lib/node/ledger/shell/prepare_proposal.rs

```diff
@@ -204,10 +357,107 @@ where
+                                bin_space_left,
+                                proposal_height =
+                                    ?self.storage.get_current_decision_height(),
+                                "Dropping decrypted tx from the current proposal",
+                            );
+                            false
+                        }
+                        AllocFailure::OverflowsBin { bin_size } => {
+                            tracing::warn!(
+                                ?tx_bytes,
+                                bin_size,
+                                proposal_height =
+                                    ?self.storage.get_current_decision_height(),
+                                "Dropping large decrypted tx from the current proposal",
+                            );
+                            true
+                        }
+                    },
+                    |()| true,
+                )
+            })
+            .collect();
+        let alloc = alloc.next_state();
+
+        (txs, alloc)
+    }
+
+    /// Builds a batch of transactions that can fit in the
+    /// remaining space of the [`BlockSpaceAllocator`].
+    fn build_remaining_batch(
+        &mut self,
+        mut alloc: BlockSpaceAllocator<FillingRemainingSpace>,
+        tx_indices: &VecIndexSet<u128>,
+        txs: Vec<TxBytes>,
+    ) -> Vec<TxBytes> {
+        get_remaining_txs(tx_indices, txs)
+            .take_while(|tx_bytes| {
+                alloc.try_alloc(&*tx_bytes).map_or_else(
+                    |status| match status {
+                        AllocFailure::Rejected { bin_space_left } => {
+                            tracing::debug!(
+                                ?tx_bytes,
+                                bin_space_left,
+                                proposal_height =
+                                    ?self.storage.get_current_decision_height(),
+                                "Dropping tx from the current proposal",
+                            );
+                            false
+                        }
+                        AllocFailure::OverflowsBin { bin_size } => {
+                            // TODO: handle tx whose size is greater
+                            // than bin size
+                            tracing::warn!(
+                                ?tx_bytes,
+                                bin_size,
+                                proposal_height =
+                                    ?self.storage.get_current_decision_height(),
+                                "Dropping large tx from the current proposal",
+                            );
+                            true
+                        }
+                    },
+                    |()| true,
+                )
+            })
             .collect()
     }
 }
 
+/// Return a list of the protocol transactions that haven't
+/// been marked for inclusion in the block, yet.
+fn get_remaining_txs(
```

Comment:
nit: could be named `get_remaining_protocol_txs` if it only relates to protocol txs? (similarly, the `tx_indices` param could be named `protocol_tx_indices`)